### PR TITLE
Remove Unverified Blocks after using them

### PR DESCRIPTION
**Pull request by vncoelho**
@erikzhang @jsolman,

Maybe we are missing blocks in the `block_cache_unverified`.

Sometimes we can have blocks there and then we receive the headers.
In this sense, we should resend them in order that they enter in the `block_cache` or are directed Persisted.

What do you think?

### CODE REVIEW

**Comment by vncoelho** [comment]
If that is the case, I am not sure if that would be the best implementation. Fell free to modify it.

**Comment by erikzhang** [comment]
I think it should not lose that block because:

https://github.com/neo-project/neo/blob/56c1ee971f374283966aeb49ac0d061df83a1559/neo/Ledger/Blockchain.cs#L316-L320

**Comment by vncoelho** [comment]
Hi @erikzhang, 
I think that you are probably right, Erik. However, I have been looking at it and I am still not sure.

The case I thought is:
`Height = 1999`
`HeaderCounter=2000`
This is kind of a normal case because nodes are both on top and with same heights and headers, the node just miss one block.

We receive block `2001` and adds to `unverified`
We receive the header 2001, 2002, 2003 and 2004.
Then, we receive the block index 2002.

Then, we receive block 2000. But I think it will not enter in the lines you mentioned. It would enter here: https://github.com/neo-project/neo/blob/56c1ee971f374283966aeb49ac0d061df83a1559/neo/Ledger/Blockchain.cs#L327-L342

Maybe we still miss block 2001.

**Comment by vncoelho** [comment]
hi @erikzhang 

It is very strange, the experimented was replicated and we got the message (The error is due to some minor problem in the design of the Add and Remove):
```cs
[11:55:10.344] chain sync: expected=120 current: 69 nodes=5
[11:55:10.432] timeout: height=70 view=0 state=Backup, ViewChanging
[11:55:10.432] request change view: height=70 view=0 nv=4 state=Backup, ViewChanging
[11:55:10.432] CheckExpectedView: view_number=5 context.ViewNumber=0 nv=5 state=Backup, ViewChanging
[11:55:11.448] chain sync: expected=121 current: 69 nodes=5
[11:55:11.458] chain sync: expected=121 current: 69 nodes=5
[11:55:11.994] persist block: 0xc632d568712f2716f616273560f6bdd22df511d1c3d4071e1ee466d889f4de13
[11:55:11.994] initialize: height=71 view=0 index=0 role=Backup
[11:55:11.996] persist block: 0xc3bdc7d7631334f407ac96643f8f885cd7157f30f1880c7799a9f4e6ff1013e6
[11:55:11.996] initialize: height=72 view=0 index=0 role=Primary

** This is strange blockUnverified.Index:{blockUnverified.Index} **

[ERROR][12/05/2018 11:55:12][Thread 0031][akka://NeoSystem/user/$a] Collection was modified; enumeration operation may not execute.
Cause: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion()
   at System.Collections.Generic.Dictionary`2.ValueCollection.Enumerator.MoveNext()
   at Neo.Ledger.Blockchain.OnNewBlock(Block block)
   at Neo.Ledger.Blockchain.OnReceive(Object message)
   at Akka.Actor.UntypedActor.Receive(Object message)
   at Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message)
   at Akka.Actor.ActorCell.ReceiveMessage(Object message)
   at Akka.Actor.ActorCell.Invoke(Envelope envelope)
[ERROR][12/05/2018 11:55:12][Thread 0022][akka://NeoSystem/user/$a] Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Persistence.LevelDB.LevelDBStore)
Cause: [akka://NeoSystem/user/$a#1273438166]: Akka.Actor.PostRestartException: Exception post restart (System.InvalidOperationException) ---> System.TypeLoadException: Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Persistence.LevelDB.LevelDBStore) ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at Neo.Ledger.Blockchain..ctor(NeoSystem system, Store store)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at System.Activator.CreateInstance(Type type, Object[] args)
   at Akka.Actor.Props.ActivatorProducer.Produce()
   at Akka.Actor.Props.NewActor()
   --- End of inner exception stack trace ---
   at Akka.Actor.Props.NewActor()
   at Akka.Actor.ActorCell.CreateNewActorInstance()
   at Akka.Actor.ActorCell.<>c__DisplayClass109_0.<NewActor>b__0()
   at Akka.Actor.ActorCell.UseThreadContext(Action action)
   at Akka.Actor.ActorCell.NewActor()
   at Akka.Actor.ActorCell.FinishRecreate(Exception cause, ActorBase failedActor)
   --- End of inner exception stack trace ---
```

The piece of code is:
```cs
            if (block.Index == Height + 1)
            {
                Block block_persist = block;
                List<Block> blocksToPersistList = new List<Block>();

                foreach (Block blockUnverified in block_cache_unverified.Values)
                {
                       if (blockUnverified.Index <= block_persist.Index)
                       {
                                Console.WriteLine("\n\n\n\n\n\n\n\n This is strange blockUnverified.Index:{blockUnverified.Index} \n\n\n\n\n\n");
                                if (blockUnverified.Verify(currentSnapshot))
                                    blocksToPersistList.Add(blockUnverified);
                                block_cache_unverified.Remove(blockUnverified.Index);
                        }
                }
```

It should not enter in that if, because the block_cache_unverified should not contain index that are lower than the one being Persisted.

**Comment by jsolman** [comment]
I believe Erik is correct. Line 319 will still pick up the block from `block_cache_unverified` at the right time. While it will hit the code you added first sometimes , that doesn’t really make adding that code necessary.  The reason it will still work without the code you are adding here is that the block_cache_unverified will never contain the very next block to be persisted, so it will always hit the code path in line 319 to re-send the blocks in `block_cache_unverified`.

Therefore the net effect of this code you are adding should just be an extra step of moving the block to verify sooner and intermediately be moved to the `block_cache` for no real adddd benefit, since it would otherwise just get picked up naturally on line 319 when persisting the block just before the first one in `block_cache_unverified`.

**Comment by vncoelho** [comment]
Perfect, @jsolman.

I agree with that, but we still have the code reaching that aforementioned line. Take a look in my last comment.

**Comment by jsolman** [comment]
My comment is saying it doesn’t matter if the code reaches the aforementioned line, it isn’t an issue if it does.

**Comment by vncoelho** [comment]
How is it not an issue if previous cached blocks are not being used?

**Comment by vncoelho** [comment]
I cleaned the code, It still need to be modified to LinkedList that you created, @jsolman.

Check if it makes sense. From my perspective, it is very strange that this `if` is activated.

**Comment by jsolman** [comment]
Hmm having extra blocks in block_cache_unverified would be a strange anomaly that shouldn’t happen. I’ll take a look to see if anything makes sense that could cause that.

**Comment by vncoelho** [comment]
Eagle eyes, @jsolman!
Thanks for the tip.

**Comment by jsolman** [comment]
Yeah, was nice chatting offline; it was an oversight that it was not deleting the blocks from unverified after using them.

File: neo/Ledger/Blockchain.cs [hunk 1]
```diff
@@ -290,7 +290,7 @@ private RelayResultReason OnNewBlock(Block block)
             {
                 Block block_persist = block;
                 List<Block> blocksToPersistList = new List<Block>();
-
+
```
Comment (jsolman) [inline neo/Ledger/Blockchain.cs]: remove extra white space

Comment (vncoelho) [reply neo/Ledger/Blockchain.cs]: EHauheauea, coincidentally we did it together.

Comment (vncoelho) [reply neo/Ledger/Blockchain.cs]: In fact, I removed the line.

**Comment by vncoelho** [comment]
I am still not sure it would delete it in the next iteration...aehauheauea
Anyway, as soon as deleted better it is.

**Pull request merged**